Sending and receiving custom headers

Encamy opened the pull request:
The PR https://github.com/alfredh/baresip/pull/425 was reworked

It was desirable to use the "X-" prefix with headers which are not defined by IETF/IANA, but not necessarily. It just removes possible collisions. But in this [RFC](https://tools.ietf.org/html/rfc6648) the "X-" prefix was deprecated. So I think that users should decide to use or not to use this prefix by themselves.

Now this feature uses list structure.

Interaction with custom headers changed:
Example of creating and sending custom headers:
```c++
int some_id = 7;
struct list *custom_hdrs;

err = custom_hdrs_alloc(&custom_hdrs);
err = custom_hdrs_add_int(custom_hdrs, "X-CALL_ID", some_id);
err = custom_hdrs_add(custom_hdrs, "X-HEADER_NAME", "VALUE");
err = ua_connect(user_agent, 0, NULL, uri, NULL, VIDMODE_OFF, custom_hdrs);

mem_deref(custom_hdrs);
```
Then the INVITE request will look like this:
![image](https://user-images.githubusercontent.com/10367317/41843847-36a5d5c4-7877-11e8-9966-b5149af23667.png)

Example of getting incoming headers:
```c++
//after creating an user agent:
ua_add_xhdr_filter(user_agent, "X-CALL_ID");

//when new call is incoming:
const struct list *hdrs;
hdrs = call_get_custom_hdrs(call); 

//Now we can interact with received headers
static int custom_header_extractor(const struct pl *name, const struct pl *val, void *arg)
{
    //do smth with "name" and "val" variables
    return 0;
}
custom_hdrs_apply(hdrs, custom_header_extractor, NULL);
```

Also a new test in baresip selftest was added

alfredh commented:
many thanks for this work.

I will review it after baresip 0.5.10 is released.

alfredh commented on include/baresip.h:
```diff
@@ -151,6 +151,22 @@ struct call  *call_find_linenum(const struct list *calls, uint32_t linenum);
+const struct list *call_get_custom_hdrs(const struct call *call);
+typedef int (custom_hdrs_h)(const struct pl *name, const struct pl *val,
+	void *arg);     //returns error code if any
+int custom_hdrs_alloc(struct list **hdrs);
```
please insert some empty lines here, between the typedef and
the function prototypes.

also please use C89 style comments (`/* xxx */`)

alfredh commented on include/baresip.h:
```diff
@@ -643,7 +659,8 @@ typedef void (ua_exit_h)(void *arg);
 		const char *from_uri, const char *uri,
-		const char *params, enum vidmode vmode);
+		const char *params, enum vidmode vmode,
+		struct list *custom_hdrs);
```
instead of breaking the `ua_connect` api, is it better to add a new function to set
the custom headers ? e.g. `ua_set_custom_hdrs()`

alfredh commented on src/call.c:
```diff
@@ -82,6 +82,7 @@ struct call {
 
 	uint32_t rtp_timeout_ms;  /**< RTP Timeout in [ms]                  */
 	uint32_t linenum;         /**< Line number from 1 to N              */
+	struct list *custom_hdrs; /**< List of custom headers if any */
```
is this for the outgoing call or the incoming call ?

Encamy replied:
This is for outgoing calls. It is using in `ua_connect`.

alfredh commented on src/call.c:
```diff
@@ -688,6 +690,20 @@ int call_alloc(struct call **callp, const struct config *cfg, struct list *lst,
 
 
+void call_set_custom_hdrs(struct call *call, struct list *hdrs)
+{
```
please check `call` parameter, return if NULL.

alfredh commented on src/call.c:
```diff
@@ -688,6 +690,20 @@ int call_alloc(struct call **callp, const struct config *cfg, struct list *lst,
+
+
+const struct list *call_get_custom_hdrs(const struct call *call)
+{
```
please check `call` parameter, return if NULL.

alfredh commented on src/custom_hdrs.c:
```diff
@@ -0,0 +1,114 @@
+#include <re.h>
```
please add a copyright header here, similar to all the other .c files

alfredh commented on src/custom_hdrs.c:
```diff
@@ -0,0 +1,114 @@
+	/*
+	list_unlink(&hdr->le);
+	hash_unlink(&hdr->he);
+	*/
```
commented out code should be removed.

alfredh commented on src/custom_hdrs.c:
```diff
@@ -0,0 +1,114 @@
+	pl_set_str(&val_pl, val);
+
+	return custom_hdrs_add_pl(hdrs, &name_pl, &val_pl);
+}
```
I am wondering if we should add formatted arguments here, similar to:

```
int sip_treplyf(struct sip_strans **stp, struct mbuf **mbp, struct sip *sip,
		const struct sip_msg *msg, bool rec_route, uint16_t scode,
		const char *reason, const char *fmt, ...)
{
	va_list ap;
	int err;

	va_start(ap, fmt);
	err = vreplyf(stp, mbp, true, sip, msg, rec_route, scode, reason,
		      fmt, ap);
	va_end(ap);

	return err;
}
```

in this case we only need one add function, which becomes very flexible and powerful.

instead of custom_hdrs_add_int you can then write:

`custom_hdrs_add(hdrs, "X-Header", "%d", 42);`

what do you think ?

alfredh commented on src/custom_hdrs.c:
```diff
@@ -0,0 +1,114 @@
+	int err;
+	struct le *le;
+	for (le = list_head(hdrs); le; le = le->next) {
+		struct sip_hdr * hdr = (struct sip_hdr *)(le->data);
```
the casting is not needed, can be removed.

alfredh commented on src/srcs.mk:
```diff
@@ -38,6 +38,7 @@ SRCS	+= timer.c
 SRCS	+= timestamp.c
 SRCS	+= ua.c
 SRCS	+= ui.c
+SRCS	+= custom_hdrs.c
```
files in alphabetical order is nice :)

alfredh commented on src/ua.c:
```diff
@@ -31,6 +31,7 @@ struct ua {
 	int af_media;                /**< Preferred Address Family for media */
 	enum presence_status my_status; /**< Presence Status                 */
 	bool catchall;               /**< Catch all inbound requests         */
+	struct list *hdr_filter;     /**< List of custom headers             */
```
is this for outgoing or incoming ?

Encamy replied:
This is for incoming calls

alfredh commented on src/ua.c:
```diff
@@ -1311,6 +1323,35 @@ static void sipsess_conn_handler(const struct sip_msg *msg, void *arg)
 	}
 
+	if (ua->hdr_filter)
+	{
```
please move curly braces to previous line, at the end:

```
if (x) {
}
```

alfredh commented on src/ua.c:
```diff
@@ -1311,6 +1323,35 @@ static void sipsess_conn_handler(const struct sip_msg *msg, void *arg)
+				NULL, NULL);
+			if (hdr_local) {
+				custom_hdrs_add_pl(hdrs, &hdr_local->name,
+					&hdr_local->val);
```
return value must be checked here

alfredh commented on src/ua.c:
```diff
@@ -1323,6 +1364,50 @@ static void sipsess_conn_handler(const struct sip_msg *msg, void *arg)
+ {
+struct ua_xhdr_filter *filter = arg;
+mem_deref(filter->hdr_name);
+}
```
this function needs indentation (1 TAB per line)

alfredh commented on src/ua.c:
```diff
@@ -1323,6 +1364,50 @@ static void sipsess_conn_handler(const struct sip_msg *msg, void *arg)
+		return ENOMEM;
+
+	}
+	strcpy(buf, hdr_name);
```
instead of these 6 lines you can use the `str_dup` function

alfredh commented:
some comments inline :)

another question is, should the custom headers also be added to the 200 OK
for incoming calls ?

Encamy commented:
Thanks for comments. I will improve code as soon as possible.
It is possible to add custom headers to response 200 OK, but I don't think that it is necessary. We just provide additional information to the client and do not expect that the client will respond to this information.

alfredh commented:
you can just add more commits to this branch.

when the review is complete I will squash and merge it master (as 1 commit).

Encamy commented:
Hello @alfredh 
Using formatted arguments is awesome idea. It is much more flexible now.
Example of usage:
```c++
custom_hdrs_add(custom_hdrs, "X-HEADER_NAME", "%s", "VALUE");
custom_hdrs_add(custom_hdrs, "X-NAME", "%d", 12)
```
Also I have removed additional parameter from `ua_connect` (api is unchanged now)
Custom headers can be applied by using `ua_set_custom_hdrs(some_headers)` before `ua_connect`.
Is it ok, that custom headers is stored in `ua.c`? Or it's better to move `static struct list *custom_hdrs` to `custom_headers.c`?

alfredh commented on src/ua.c:
```diff
@@ -81,6 +87,7 @@ static struct {
 #endif
 };
 
+static struct list *custom_hdrs;
```
it is better to move this to `struct ua` so that it becomes re-entrant.

the long term goal of baresip is to make it 100% re-entrant with no static/global data.

alfredh commented on test/call.c:
```diff
@@ -969,3 +980,75 @@ int test_call_mediaenc(void)
 }
+
+
+int test_call_custom_headers(void) {
```
the curly brace should be moved to the next line.

functions are written like this:

```
void function(void)
{
}
```

alfredh commented on test/call.c:
```diff
@@ -969,3 +980,75 @@ int test_call_mediaenc(void)
+	/* Make a call from A to B
+	 * with some custom headers in INVITE message */
+	int some_id = 7;
+	struct list *custom_hdrs;
```
declaration of variables should be moved to the beginning of the function.

alfredh commented on test/call.c:
```diff
@@ -969,3 +980,75 @@ int test_call_mediaenc(void)
+	/* run main-loop with timeout, wait for events */
+	err = re_main_timeout(5000);
+
+	bool headers_matched = true;
```
declaration of variables should be moved to the beginning of the function.

alfredh commented on src/custom_hdrs.c:
```diff
@@ -0,0 +1,95 @@
+	list_init(h);
+	*hdrs = h;
+	return 0;
+}
```
I am wondering if it is better to use a compound list instead of allocated list.

Most of the code in baresip and libre using a list is declared like this:

```
struct foo {
    struct list my_list;
};
```

this way there is no need for an allocation. The list is empty by default.

To initialize the list:

`list_init();`

To destroy the list members:

`list_flush();`

To check if the list is empty or have any members:

`list_isempty()` or `list.head != NULL`

if you write the code like this you will find that it becomes simpler.
but it is not possible to use `mem_ref` on it, so you have to copy each element.

alfredh commented on src/custom_hdrs.c:
```diff
@@ -0,0 +1,95 @@
+	err = re_vsdprintf(&value, fmt, ap);
+	va_end(ap);
+
+	struct sip_hdr *hdr;
```
please move variable declaration to the beginning of the function

alfredh commented on src/custom_hdrs.c:
```diff
@@ -0,0 +1,95 @@
+	return(err);
+}
+
+int custom_hdrs_apply(const struct list *hdrs,
```
the distance between functions should be 2 empty lines.

alfredh commented on src/custom_hdrs.c:
```diff
@@ -0,0 +1,95 @@
+		}
+	}
+	return 0;
+}
```
can you add some whitespace here, at least empty line before and after
the for loop. the code becomes a bit more readable.

alfredh commented on include/baresip.h:
```diff
@@ -151,6 +151,20 @@ struct call  *call_find_linenum(const struct list *calls, uint32_t linenum);
+/*
+* Custom headers
+*/
+const struct list *call_get_custom_hdrs(const struct call *call);
```
perhaps move this prototype to the `struct call` block since it is accessing `call` ?

alfredh commented:
thanks for the update. it looks better now.

please have a second look at the formatting. all functions should have 2 empty
lines between them. The indentation is TAB. also insert some extra whitespace
to make the code more readable :)

Alfred

Encamy commented:
I have fixed code style and moved list of custom headers to ua struct.
It is possible to change lists from allocated list to compound list. But I have some troubles with this task. I will change them if it is necessary. Should I do that?

alfredh commented on src/custom_hdrs.c:
```diff
@@ -0,0 +1,105 @@
+		goto error;
+
+	pl_set_str(&temp_pl, name);
+	pl_dup(&hdr->name, &temp_pl);
```
return value of `pl_dup` must be checked.

alfredh commented on src/custom_hdrs.c:
```diff
@@ -0,0 +1,105 @@
+		err = h(&hdr->name, &hdr->val, arg);
+		if (err) {
+			return err;
+		}
```
no need for curly braces here, since it has only 1 line.

alfredh commented on src/ua.c:
```diff
@@ -549,6 +556,8 @@ static void ua_destructor(void *arg)
 		sip_close(uag.sip, false);
 	}
+
+	mem_deref(ua->hdr_filter);
```
also `custom_hdrs` must be derefed in the destructor

alfredh commented on src/ua.c:
```diff
@@ -833,13 +842,18 @@ int ua_connect(struct ua *ua, struct call **callp,
 		*callp = call;
 
+	ua_set_custom_hdrs(ua, NULL);
+
```
I think we should clear the list of custom headers here, this should be done by the application.

alfredh commented on src/ua.c:
```diff
@@ -1311,6 +1325,35 @@ static void sipsess_conn_handler(const struct sip_msg *msg, void *arg)
+
+			le = le->next;
+		    hdr_local = sip_msg_xhdr_apply(msg, true, filter->hdr_name,
+				NULL, NULL);
```
you can also use `sip_msg_xhdr` here, it is perhaps a bit simpler.

alfredh commented on src/ua.c:
```diff
@@ -1323,6 +1366,47 @@ static void sipsess_conn_handler(const struct sip_msg *msg, void *arg)
+
+	char *buf;
+	if (str_dup(&buf, hdr_name))
+		return ENOMEM;
```
instead of using a temp buf, you can instead dup the string
directly into `filter->hdr_name`

in case of errors, the destructor will deref the string.

alfredh commented on src/ua.c:
```diff
@@ -1965,3 +2049,9 @@ int uag_set_extra_params(const char *eprm)
+void ua_set_custom_hdrs(struct ua *ua, struct list *custom_headers)
+{
+	ua->custom_hdrs = custom_headers;
+}
```
you should use mem_deref and mem_ref here.

the data in struct ua should not point to objects from the application.
the application object might go out of scope, which will make the
`ua->custom_hdrs` pointer point to invalid data (and perhaps a crash).

Encamy replied:
Sorry, I can't understand how exactly to use `mem_deref` and `mem_ref`.
Should this look something like the following?
```c++
mem_ref(custom_headers);
ua->custom_hdrs = custom_headers;
mem_deref(custom_headers);
```

alfredh commented on src/ua.c:
```diff
@@ -31,6 +31,8 @@ struct ua {
 	int af_media;                /**< Preferred Address Family for media */
 	enum presence_status my_status; /**< Presence Status                 */
 	bool catchall;               /**< Catch all inbound requests         */
+	struct list *hdr_filter;     /**< Filter for incoming headers        */
```
I think the filter can be converted to a compound list, that needs no allocation.

I have made an example of it here:

https://gist.github.com/alfredh/2577a5dadbde35a9ba47d96696ddb31d

Encamy commented:
Thanks for a review. I have made some changes to the code.
Also I have tried to replace every allocated lists with compound lists. Code became more complicated. The most suitable method for copying all data is to add pairs `name:value` from one list to another. We can't just copy the memory of one header and paste to other because `mem_deref(hdr)` will break the application.
For example:
```c++
struct le *le;
LIST_FOREACH(&hdrs, le) {
	struct sip_hdr *hdr = le->data;
	custom_hdrs_add(&call->custom_hdrs, hdr->name.p, "%r", &hdr->val);
}
```
I don't think that it is good decision. May be I'm missing something? Should I replace other allocated lists with compound or we can leave it as it is?

Encamy commented:
I have found out the problem. This version is using only compound lists.

alfredh commented:
this looks very good. I have a small comment, see inline.

alfredh commented on src/ua.c:
```diff
@@ -1965,3 +2035,18 @@ int uag_set_extra_params(const char *eprm)
 }
+
+
+void ua_set_custom_hdrs(struct ua *ua, struct list custom_headers)
```
the custom_headers parameter should be of type pointer to a const struct list:

`const struct list *custom_headers`

it is not recommended to pass structs like this as function parameters.

alfredh commented on src/call.c:
```diff
@@ -688,6 +691,33 @@ int call_alloc(struct call **callp, const struct config *cfg, struct list *lst,
+	LIST_FOREACH(&hdrs, le) {
+		struct sip_hdr *hdr = le->data;
+		char *buf = NULL;
+		re_sdprintf(&buf, "%r", &hdr->name);
```
check return value, return if errors.

alfredh commented on src/call.c:
```diff
@@ -688,6 +691,33 @@ int call_alloc(struct call **callp, const struct config *cfg, struct list *lst,
+		struct sip_hdr *hdr = le->data;
+		char *buf = NULL;
+		re_sdprintf(&buf, "%r", &hdr->name);
+		custom_hdrs_add(&call->custom_hdrs, buf, "%r", &hdr->val);
```
check return value, return if errors.

Encamy commented:
Thank you. I have fixed these shortcomings.

Encamy commented:
Thank you for all your reviews and for help
